game-core/src/main/java/games/strategy/triplea/delegate/ScrambleLogic.java

```diff
@@ -0,0 +1,223 @@
+import games.strategy.engine.data.Territory;
+import games.strategy.engine.data.Unit;
+import games.strategy.engine.data.UnitCollection;
+import games.strategy.engine.data.UnitType;
+import games.strategy.triplea.Properties;
+import games.strategy.triplea.TripleAUnit;
+import games.strategy.triplea.attachments.UnitAttachment;
+import games.strategy.triplea.delegate.IBattle.BattleType;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Predicate;
+import lombok.Getter;
+import org.triplea.java.PredicateBuilder;
+import org.triplea.java.collections.CollectionUtils;
+import org.triplea.util.Tuple;
+
+/**
+ * ScrambeLogic encapsulates the logic for finding possible units that can scramble to defend a
+ * given territory.
+ */
+public class ScrambleLogic {
+  private final GameData data;
+  private final PlayerId player;
+  private final Set<Territory> territoriesWithBattles;
+  private BattleTracker battleTracker;
+  @Getter private final Predicate<Unit> airbaseThatCanScramblePredicate;
+  private final Predicate<Territory> canScrambleFromPredicate;
+  private final int maxScrambleDistance;
+
+  public ScrambleLogic(
+      final GameData data,
+      final PlayerId player,
+      final Set<Territory> territoriesWithBattles,
+      final BattleTracker battleTracker) {
+    this.data = data;
+    this.player = player;
+    this.territoriesWithBattles = territoriesWithBattles;
+    this.battleTracker = battleTracker;
+    this.airbaseThatCanScramblePredicate =
+        Matches.unitIsEnemyOf(data, player)
+            .and(Matches.unitIsAirBase())
+            .and(Matches.unitIsNotDisabled())
+            .and(Matches.unitIsBeingTransported().negate());
+    this.canScrambleFromPredicate =
+        PredicateBuilder.of(Matches.territoryIsWater().or(Matches.isTerritoryEnemy(player, data)))
+            .and(
+                Matches.territoryHasUnitsThatMatch(
+                    Matches.unitCanScramble()
+                        .and(Matches.unitIsEnemyOf(data, player))
+                        .and(Matches.unitIsNotDisabled())))
+            .and(Matches.territoryHasUnitsThatMatch(airbaseThatCanScramblePredicate))
+            .andIf(Properties.getScrambleFromIslandOnly(data), Matches.territoryIsIsland())
+            .build();
+    this.maxScrambleDistance = computeMaxScrambleDistance(data);
+  }
+
+  private static int computeMaxScrambleDistance(final GameData data) {
+    int maxScrambleDistance = 0;
+    for (final UnitType unitType : data.getUnitTypeList()) {
+      final UnitAttachment ua = UnitAttachment.get(unitType);
+      if (ua.getCanScramble() && maxScrambleDistance < ua.getMaxScrambleDistance()) {
+        maxScrambleDistance = ua.getMaxScrambleDistance();
+      }
+    }
+    return maxScrambleDistance;
+  }
+
+  public ScrambleLogic(final GameData data, final PlayerId player, final Territory territory) {
```

Comment:
minor: moving this constructor above the first constructor would minimize vertical distances. _Clean Code_ emphasizes with depth-first ordering that you want readers, to the extent possible, to be able to read code with one vertical scan from top to bottom. With the re-ordering, there would be no need to jump back up the page to lookup up the `this(..)` reference.

I'll update the convention to give that as a bullet point:
https://github.com/triplea-game/triplea/wiki/Java-Code-Conventions#depth-first-method-ordering

Reply:
Done.